articles/contributing/style-guide.md

```diff
@@ -0,0 +1,451 @@
+***
+
+### Variants ###
+
+Though this limitation may not persist in future versions of Q#, it is presently the case that there will often be groups of related operations or functions that are distinguished by which functors their inputs support, or by the concrete types of their arguments.
+These groups can be distinguished by using the same root name, followed by one or two letters that indicate its variant.
+
+| Suffix | Meaning |
+|--------|---------|
+| `A` | Input expected to support `Adjoint` |
+| `C` | Input expected to support `Controlled` |
+| `CA` | Input expected to support `Controlled` and `Adjoint` |
+| `I` | Input or inputs are of type `Int` |
+| `D` | Input or inputs are of type `Double` |
+| `L` | Input or inputs are of type `BigInt` |
+| `BE` | Input or inputs are of type `BigEndian` |
+| `LE` | Input or inputs are of type `LittleEndian` |
+
+# [Guidance](#tab/guidance)
+
+We suggest:
+
+- If a function or operation is not related to any similar functions or operations by the types and functor support of their inputs, do not use a suffix.
+- If a function or operation is related to any similar functions or operations by the types and functor support of their inputs, use suffixes as in the table above to distinguish variants.
+
+# [Examples](#tab/examples)
+
+***
+
+### Arguments and Variables ###
+
+A key goal of the Q# code for a function or operation is for it to be easily read and understood.
+Similarly, the names of inputs and type arguments should communicate how a function or argument will be used once provided.
+
+
+# [Guidance](#tab/guidance) 
+
+We suggest:
+
+- For all variable and input names, use `pascalCase` in strong preference to `CamelCase`, `snake_case`, or `ANGRY_CASE`.
+- Input names should be descriptive; avoid one or two letter names where possible.
+- Operations and functions accepting exactly one type argument should denote that type argument by `T` when its role is obvious.
+- If a function or operation takes multiple type arguments, or if the role of a single type argument is not obvious, consider using a short capitalized word prefaced by `T` (e.g.: `TOutput`) for each type.
+- Do not include type names in argument and variable names unless absolutely necessary; this information can and should be provided by your development environment.
+- Denote scalar types by their literal names (`flagQubit`), and array types by a plural (`measResults`).
+  For arrays of qubits in particular, consider denoting such types by `Register` where the name refers to a sequence of qubits that are closely related in some way.
+- Variables used as indices into arrays should begin with `idx` and should be singular (e.g.: `things[idxThing]`).
+  In particular, strongly avoid using single-letter variable names as indices; consider using `idx` at a minimum.
+- Variables used to hold lengths of arrays should begin with `n` and should be pluralized (e.g.: `nThings`).
+
+# [Examples](#tab/examples)
+
+***
+
+## Input Conventions ##
```

Comment:
It may be helpful to also think about and formulate conventions regarding the "aggregation" of arguments (i.e. nesting)?

Reply:
Good point. I think I'll defer on that for now, though, and will file it as an issue for future improvement if that's OK?